kernel/src/expressions/scalars.rs

```diff
@@ -66,9 +60,26 @@ pub struct ArrayData {
 }
 
 impl ArrayData {
-    pub fn new(tpe: ArrayType, elements: impl IntoIterator<Item = impl Into<Scalar>>) -> Self {
-        let elements = elements.into_iter().map(Into::into).collect();
-        Self { tpe, elements }
+    pub fn try_new(
+        tpe: ArrayType,
+        elements: impl IntoIterator<Item = impl Into<Scalar>>,
+    ) -> DeltaResult<Self> {
+        let elements = elements
+            .into_iter()
+            .map(Into::into)
+            .map(|v| {
+                if *tpe.element_type() == v.data_type() {
+                    Ok(v)
+                } else {
+                    Err(Error::Generic(format!(
+                        "Array scalar type mismatch: expected {}, got {}",
+                        tpe.element_type(),
+                        v.data_type()
+                    )))
+                }
+            })
+            .collect::<Result<Vec<_>, _>>()?;
```

Comment:
nit:
```suggestion
            .try_collect()?;
```
(again below)